stdlib/public/SDK/Foundation/NSFastEnumeration.swift

```diff
@@ -16,63 +16,35 @@
 //===----------------------------------------------------------------------===//
 // Fast enumeration
 //===----------------------------------------------------------------------===//
-
-// NB: This is a class because fast enumeration passes around interior pointers
-// to the enumeration state, so the state cannot be moved in memory. We will
-// probably need to implement fast enumeration in the compiler as a primitive
-// to implement it both correctly and efficiently.
-final public class NSFastEnumerationIterator : IteratorProtocol {
-  var enumerable: NSFastEnumeration
-  var state: [NSFastEnumerationState]
-  var n: Int
-  var count: Int
-
-  /// Size of ObjectsBuffer, in ids.
-  static var STACK_BUF_SIZE: Int { return 4 }
-
-  var objects: [Unmanaged<AnyObject>?]
-
-  public func next() -> Any? {
-    if n == count {
-      // FIXME: Is this check necessary before refresh()?
-      if count == 0 { return nil }
-      refresh()
-      if count == 0 { return nil }
+public struct NSFastEnumerationIterator : IteratorProtocol {
+    var enumerable: NSFastEnumeration
+    var objects: (Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?, Unmanaged<AnyObject>?) = (nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil)
+    var state = NSFastEnumerationState(state: 0, itemsPtr: nil, mutationsPtr: _fastEnumerationStorageMutationsPtr, extra: (0, 0, 0, 0, 0))
+    var index = 0
+    var count = 0
+    
+    public init(_ enumerable: NSFastEnumeration) {
+        self.enumerable = enumerable
     }
-    let next: Any = state[0].itemsPtr![n]!
-    n += 1
-    return next
-  }
-
-  func refresh() {
-    _sanityCheck(objects.count > 0)
-    n = 0
-    objects.withUnsafeMutableBufferPointer {
-      count = enumerable.countByEnumerating(
-        with: &state,
-        objects: AutoreleasingUnsafeMutablePointer($0.baseAddress!),
-        count: $0.count)
+    
+    public mutating func next() -> Any? {
+        if index + 1 > count {
+            index = 0
+            count = withUnsafeMutablePointer(to: &objects) {
+                return enumerable.countByEnumerating(with: &state, objects: AutoreleasingUnsafeMutablePointer($0), count: 16)
```

Comment:
```swift
public mutating func next() -> Any? {
        if index + 1 > count {
            index = 0
            count = withUnsafeMutablePointer(to: &objects) {
                let buffer = AutoreleasingUnsafeMutablePointer<AnyObject?>($0)
                let result = enumerable.countByEnumerating(with: &state, objects: buffer, count: 16)
                if state.itemsPtr == buffer {
                    // Most cocoa classes will emit their own inner pointer buffers instead of traversing this path.
                    useObjectsBuffer = true
                } else {
                    // this is the common case
                    useObjectsBuffer = false
                }
                return result
            }
            if count == 0 { return nil }
        }
        defer { index += 1 }
        if !useObjectsBuffer {
            return state.itemsPtr![index]
        } else {
            switch index {
            case 0: return objects.0!.takeUnretainedValue()
            case 1: return objects.1!.takeUnretainedValue()
            case 2: return objects.2!.takeUnretainedValue()
            case 3: return objects.3!.takeUnretainedValue()
            case 4: return objects.4!.takeUnretainedValue()
            case 5: return objects.5!.takeUnretainedValue()
            case 6: return objects.6!.takeUnretainedValue()
            case 7: return objects.7!.takeUnretainedValue()
            case 8: return objects.8!.takeUnretainedValue()
            case 9: return objects.9!.takeUnretainedValue()
            case 10: return objects.10!.takeUnretainedValue()
            case 11: return objects.11!.takeUnretainedValue()
            case 12: return objects.12!.takeUnretainedValue()
            case 13: return objects.13!.takeUnretainedValue()
            case 14: return objects.14!.takeUnretainedValue()
            case 15: return objects.15!.takeUnretainedValue()
            default: fatalError("Access beyond storage buffer")
            }
        }
    }
```

How bout that? Does that address the concern of stack use after valid contexts?